modules/packages/LinearAlgebra.chpl

```diff
@@ -2500,6 +2517,410 @@ proc cosm(A: []) throws {
+private proc absSum(A: [?ADom], axis=0) throws where isCSDom(ADom) {
+  if (axis != 0 && axis != 1) {
+    throw new LinearAlgebraError("absSum(): axis=0 or axis=1 expected");
+  }
+
+  var (n,t) = ADom.shape;
+  if axis == 0 {
+    // axis=0 sums along column
+    var b: [0..<t] A.eltType;
+    forall (i,j) in ADom with (+ reduce b) {
+      b[j] += abs(A[i,j]);
+    }
+    return b;
+  }
+  else {
+    // axis=1 sums along rows
+    var b: [0..<n] A.eltType;
+    forall (i,j) in ADom with (+ reduce b) {
+      b[i] += abs(A[i,j]);
+    }
+    return b;
+  }
+}
+
+// Gets max along specified axes. Method for CSR matrix.
+private proc maxAlongAxis(A: [?ADom], axis=0) throws where isCSDom(ADom) {
+  if (axis != 0 && axis != 1) {
+    throw new LinearAlgebraError("absSum(): axis=0 or axis=1 expected");
+  }
+
+  var (n,t) = ADom.shape;
+  if axis == 0 {
+    // axis=0 sums along column
+    var b: [0..<t] A.eltType;
+    forall (i,j) in ADom with (max reduce b) {
+      b[j] = max(b[j], A[i,j]);
+    }
+    return b;
+  }
+  else {
+    // axis=1 sums along rows
+    var b: [0..<n] A.eltType;
+    forall (i,j) in ADom with (max reduce b) {
+      b[i] = max(b[i], A[i,j]);
+    }
+    return b;
+  }
+}
+
+// Gets max along specified axes. Method for Rectangular matrix.
+private proc maxAlongAxis(A: [?ADom], axis=0) throws where !isCSDom(ADom) {
+  if (axis != 0 && axis != 1) {
+    throw new LinearAlgebraError("absSum(): axis=0 or axis=1 expected");
+  }
+
+  var (n,t) = ADom.shape;
+  if axis == 0 {
+    // axis=0 sums along column
+    var b: [0..<t] A.eltType;
+    forall i in 0..<t {
+      b[i] = max reduce A[.., i];
+    }
+    return b;
+  }
+  else {
+    // axis=1 sums along rows
+    var b: [0..<n] A.eltType;
+    forall i in 0..<n {
+      b[i] = max reduce A[i, ..];
+    }
+    return b;
+  }
+}
+
+// Given the Domain and element-type, method
+// returns a Matrix of ones.
+private proc ones(Dom: domain(2), type eltType=real) {
+  var A: [Dom] eltType = 1 : eltType;
+  return A;
+}
+
+
+// Given the Domain and element-type, method
+// returns a Matrix of ones.
+private proc zeros(Dom: domain(2), type eltType=real) {
+  var A: [Dom] eltType = 0 : eltType;
+  return A;
+}
+
+
+// Returns an index's unit vector of size n.
+private proc elementaryVector(n,ind) {
+  var v: [0..<n] real;
+  v[ind] = 1;
+  return v;
+}
+
+record Cmp {
+  param desc;
+  var A;
+}
+
+proc Cmp.compare(a, b) {
+  if desc then {
+    return -(A[a] - A[b]);
+  }
+  else {
+    return A[a] - A[b];
+  }
+}
+
+// Given an array A, returns indexes
+// corresponding to the sorted ordering of A.
+private proc argsort(A: [], param desc=false) {
+  use Sort;
+  var cmp = new Cmp(desc, A);
+  var B = [ii in A.domain] ii;
+  sort(B, comparator=cmp);
+  return B;
+}
+
+// Method returns false if there is atleast one
+// column in X which isn't parallel to a column
+// in Y. Else the method returns true.
+private proc everyColOfXParallelToColOfY(X: [?xD], Y: [?yD]){
```

Comment:
I think we discussed this, but I forgot. When you say "parallel" here, does it mean equal? Then why is this not just `X == Y`?

Reply:
Yes, it means equal. When `X == Y` is used, it throws `error: iterator or promoted expression iterator used in if or while condition` error.

Reply:
Ah, right. Because vector `==` returns element-wise equality in another vector. You need `&& reduce X==Y`. I think that should work.